src/base/bittorrent/session.cpp

```diff
@@ -671,124 +638,134 @@ QString Session::torrentTempPath(const TorrentInfo &torrentInfo) const
+QStringList Session::expandCategoryName(const QString &categoryName)
 {
     QStringList result;
-    if (!isValidCategoryName(category))
-        return result;
 
     int index = 0;
-    while ((index = category.indexOf('/', index)) >= 0) {
-        result << category.left(index);
+    while ((index = categoryName.indexOf('/', index)) >= 0) {
+        result << categoryName.left(index);
         ++index;
     }
-    result << category;
+    result << categoryName;
 
     return result;
 }
 
-QStringList Session::categories() const
+const CategoryMap Session::categories() const
 {
-    return m_categories.keys();
+    return m_categories;
 }
 
 QString Session::categorySavePath(const QString &categoryName) const
 {
-    QString basePath = m_defaultSavePath;
-    if (categoryName.isEmpty()) return basePath;
+    if (categoryName.isEmpty())
+        return defaultSavePath();
 
-    QString path = m_categories.value(categoryName);
-    if (path.isEmpty()) // use implicit save path
-        path = Utils::Fs::toValidFileSystemName(categoryName, true);
+    QString path {Utils::Fs::toValidFileSystemName(categoryName, true)};
+    TorrentCategory *const category {findCategory(categoryName)};
+    if (category && !category->savePath().isEmpty()) // use explicit save path
+        path = category->savePath();
 
     if (!QDir::isAbsolutePath(path))
-        path.prepend(basePath);
+        path = QDir(defaultSavePath()).absoluteFilePath(path);
 
     return normalizeSavePath(path);
 }
 
-bool Session::addCategory(const QString &name, const QString &savePath)
+TorrentCategory *Session::findCategory(const QString &categoryName) const
 {
-    if (name.isEmpty()) return false;
-    if (!isValidCategoryName(name) || m_categories.contains(name))
-        return false;
+    return m_categories.value(TorrentCategory::cleanCategoryName(categoryName));
+}
+
+TorrentCategory *Session::getCategory(const QString &categoryName)
+{
+    const QString validCategoryName {TorrentCategory::cleanCategoryName(categoryName)};
+    TorrentCategory *category = m_categories.value(validCategoryName);
+    if (category) return category;
+
+    return createCategory_impl(validCategoryName);
+}
+
+TorrentCategory *Session::createCategory(const QString &categoryName)
+{
+    return createCategory_impl(TorrentCategory::cleanCategoryName(categoryName));
+}
+
+TorrentCategory *Session::createCategory_impl(const QString &validCategoryName)
+{
+    if (validCategoryName.isEmpty())
+        return nullptr;
+
+    if (m_categories.contains(validCategoryName))
+        return nullptr;
 
     if (isSubcategoriesEnabled()) {
-        foreach (const QString &parent, expandCategory(name)) {
-            if ((parent != name) && !m_categories.contains(parent)) {
-                m_categories[parent] = "";
-                emit categoryAdded(parent);
+        for (const QString &parentCategoryName: expandCategoryName(validCategoryName)) {
```

Comment:
I suppose you should add space before the colon

Reply:
We still don't have any rules about it (as well as using colon in class declaration) and we have different usage in our code. I don't care how it will be used, but let's "legalize" it first. Now it's just a subjective opinion...

Reply:
>I don't care how it will be used

I do have a preference but I can adapt to others.

>but let's "legalize" it first

OK, so... lets vote?
I vote for `for (auto i : vector)` form.

Reply:
As I wrote above, I have no preferences in this case.
>I vote for for (auto i : vector) form.

Then the class declaration should be `class ClassName : public BaseClass`, right?

Reply:
>Then the class declaration should be class ClassName : public BaseClass, right?

I agree with that form.

Reply:
Can you create PR for these changes into CODING_GUIDELINES.md?

Reply:
>Can you create PR for these changes into CODING_GUIDELINES.md?

see PR #7385.